onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlSTFT.h

```diff
@@ -0,0 +1,610 @@
+            barriers[barrierCount++] = CD3DX12_RESOURCE_BARRIER::Transition(windowResource.Get(), D3D12_RESOURCE_STATE_COMMON, D3D12_RESOURCE_STATE_UNORDERED_ACCESS);
+            inputBindingsCount++;
+        }
+
+        m_framingOperator.bindingTable->BindInputs(inputBindingsCount, inputBindings.data());
+
+        DML_BUFFER_BINDING outputBuffer = {};
+        outputBuffer.Buffer = outputResource;
+        outputBuffer.SizeInBytes = m_framingOperator.outputBufferSizeInBytes;
+        DML_BINDING_DESC outputBinding = { DML_BINDING_TYPE_BUFFER, &outputBuffer };
+        barriers[barrierCount++] = CD3DX12_RESOURCE_BARRIER::Transition(outputResource, D3D12_RESOURCE_STATE_COMMON, D3D12_RESOURCE_STATE_UNORDERED_ACCESS);
+
+        m_framingOperator.bindingTable->BindOutputs(1, &outputBinding);
+
+        ComPtr<ID3D12Resource> tempBuffer;
+        auto tempBufferSize = bindingProps.TemporaryResourceSize;
+        if (tempBufferSize > 0)
+        {
+            ORT_THROW_IF_FAILED(context->AllocateTemporaryData(tempBufferSize, &tempBuffer));
+
+            DML_BUFFER_BINDING bufferBinding = { tempBuffer.Get(), 0, tempBufferSize };
+            DML_BINDING_DESC bindingDesc = { DML_BINDING_TYPE_BUFFER, &bufferBinding };
+            m_framingOperator.bindingTable->BindTemporaryResource(&bindingDesc);
+        }
+
+        auto persistentBufferSize = bindingProps.PersistentResourceSize;
+        if (persistentBufferSize > 0)
+        {
+            DML_BUFFER_BINDING bufferBinding = { m_framingOperator.persistentResource.Get(), 0, persistentBufferSize };
+            DML_BINDING_DESC bindingDesc = { DML_BINDING_TYPE_BUFFER, &bufferBinding };
+            m_framingOperator.bindingTable->BindPersistentResource(&bindingDesc);
+        }
+
+        // Transition resources COMMON -> UAV
+        commandList->ResourceBarrier(barrierCount, barriers.data());
+
+        m_framingOperator.commandRecorder->RecordDispatch(
+            commandList,
+            m_framingOperator.op.Get(),
+            m_framingOperator.bindingTable.Get()
+        );
+
+        // Transition resources UAV -> COMMON
+        for (uint32_t barrierIndex = 0; barrierIndex < barrierCount; barrierIndex++)
+        {
+            std::swap(barriers[barrierIndex].Transition.StateBefore, barriers[barrierIndex].Transition.StateAfter);
+        }
+
+        commandList->ResourceBarrier(barrierCount, barriers.data());
+    }
+};
+
+struct STFTShapeInferrer : public WRL::Base<IMLOperatorShapeInferrer>
+{
+    STDMETHOD(InferOutputShapes)(IMLOperatorShapeInferenceContext* context) noexcept
+    {
+        try
+        {
+            ComPtr<IMLOperatorShapeInferenceContextPrivate> contextPrivate;
+            ORT_THROW_IF_FAILED(context->QueryInterface(IID_PPV_ARGS(&contextPrivate)));
+
+            DmlSTFTShapeInferenceParams info(contextPrivate.Get(), contextPrivate.Get());
+
+            std::array<uint32_t, 4> outputDims = { info.batchSize, info.frameCount, info.frameDftElementCount, 2 };
```

Comment:
> 2 [](http://example.com/codeflow?start=110&length=2)

Hmm, is this trailing 2 for the real and complex parts? If so, how does that correlate to the statement above, which says the trailing dimension must equal 1?

```
            if (rank == 3)
            {
                ML_CHECK_VALID_ARGUMENT(dims[2] == 1, "DML STFT only accepts real-valued input.");
            }
```

Reply:
Oh wait, that's for the input, not output. Resolving.